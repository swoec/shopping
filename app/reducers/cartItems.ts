const cartItems = (state = [], action) => {
  switch (action.type) {
    case "ADD_TO_CART":
      return converter(state, action);

    case "REMOVE_FROM_CART":
      return state.filter((cartItem) => cartItem.id !== action.payload.id);
  }

  return state;
};

function converter(state, action) {
  const productExists = state.find((item) => item.id === action.payload.id);
  if (productExists) {
    state.map((item) => {
      item.id === action.payload.id ? (item.quantity += 1) : item;
    });
  } else {
    action.payload.quantity = 1;
    state = [...state, action.payload];
  }

  return state;
}

export default cartItems;
