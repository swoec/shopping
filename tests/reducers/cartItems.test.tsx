import * as React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import cartItems from "../../app/reducers/cartItems";

const middlewares = [];
const mockStore = configureStore(middlewares);

describe("Dashboard", () => {
  it("add to cart correctly", () => {
    const initialState = { id: 1, name: "Sledgehammer", price: 125.75 };
    const store = mockStore(initialState);

    const product = [{ id: 1, name: "Sledgehammer", price: 125.75 }];
    let action = { type: "ADD_TO_CART", payload: product };

    const newState = cartItems([initialState], action);

    expect(newState).toMatchSnapshot();
    // Sending the same id should clean it
    expect(cartItems(newState, action)).toMatchSnapshot();
  });

  it("delete to cart correctly", () => {
    const initialState = {};
    const store = mockStore(initialState);
    const product = [
      { id: 1, name: "Sledgehammer", price: 125.75, quantity: 1 },
    ];

    let action = { type: "REMOVE_FROM_CART", payload: product };

    const newState = cartItems([initialState], action);

    expect(newState).toMatchSnapshot();
    // Sending the same id should clean it
    expect(cartItems(newState, action)).toMatchSnapshot();
  });
});
