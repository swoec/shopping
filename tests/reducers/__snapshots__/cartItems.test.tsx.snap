// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard add to cart correctly 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Sledgehammer",
    "price": 125.75,
  },
  Array [
    Object {
      "id": 1,
      "name": "Sledgehammer",
      "price": 125.75,
    },
  ],
]
`;

exports[`Dashboard add to cart correctly 2`] = `
Array [
  Object {
    "id": 1,
    "name": "Sledgehammer",
    "price": 125.75,
  },
  Array [
    Object {
      "id": 1,
      "name": "Sledgehammer",
      "price": 125.75,
    },
  ],
]
`;

exports[`Dashboard delete to cart correctly 1`] = `Array []`;

exports[`Dashboard delete to cart correctly 2`] = `Array []`;
